lhotse/array.py

```diff
@@ -0,0 +1,181 @@
+import decimal
+from dataclasses import dataclass
+from math import isclose
+from typing import List, Optional
+
+import numpy as np
+
+from lhotse import FeaturesWriter
+from lhotse.features.io import get_reader
+from lhotse.utils import Seconds, asdict_nonull, rich_exception_info
+
+
+@dataclass
+class Array:
+    """ """
+
+    # Storage type defines which features reader type should be instantiated
+    # e.g. 'lilcom_files', 'numpy_files', 'lilcom_hdf5'
+    storage_type: str
+
+    # Storage path is either the path to some kind of archive (like HDF5 file) or a path
+    # to a directory holding files with feature matrices (exact semantics depend on storage_type).
+    storage_path: str
+
+    # Storage key is either the key used to retrieve an array from an archive like HDF5,
+    # or the name of the file in a directory (exact semantics depend on the storage_type).
+    storage_key: str
+
+    # Shape of the array once loaded into memory.
+    shape: List[int]
+
+    @property
+    def ndim(self) -> int:
+        return len(self.shape)
+
+    def to_dict(self) -> dict:
+        return asdict_nonull(self)
+
+    @classmethod
+    def from_dict(cls, data: dict) -> "Array":
+        return cls(**data)
+
+    @classmethod
+    def store(cls, key: str, value: np.ndarray, writer: FeaturesWriter) -> "Array":
+        storage_key = writer.write(key, value)
+        return Array(
+            storage_type=writer.name,
+            storage_path=str(writer.storage_path),
+            storage_key=storage_key,
+            shape=list(value.shape),
+        )
+
+    @rich_exception_info
+    def load(self) -> np.ndarray:
+
+        # noinspection PyArgumentList
+        storage = get_reader(self.storage_type)(self.storage_path)
+
+        # Load and return the array from the storage
+        return storage.read(self.storage_key)
+
+
+@dataclass
+class TemporalArray:
+    """ """
+
+    # Manifest describing the base array.
+    array: Array
+
+    # Indicates which dim corresponds to the time dimension:
+    # e.g., PCM audio samples indexes, feature frame indexes, chunks indexes, etc.
+    temporal_dim: int
+
+    # The time interval (in seconds, or fraction of a second) between the start timestamps
+    # of consecutive frames. Only defined when ``temporal_axis`` is not ``None``.
+    frame_shift: Seconds
+
+    # Information about the time range of the features.
+    # We only need to specify start, as duration can be computed from
+    # the shape, temporal_axis, and frame_shift.
+    start: Seconds
+
+    @property
+    def shape(self):
+        return self.array.shape
+
+    @property
+    def ndim(self) -> int:
+        return len(self.shape)
+
+    @property
+    def duration(self) -> Optional[Seconds]:
```

Comment:
The return type is `Optional[Seconds]`. In which case will it return a `None`?

Reply:
Left-over from a previous draft. Will fix.